src/audio_device.c

```diff
@@ -37,131 +33,181 @@
+
+    // used only by output devices:
+    uint32_t source;
+    uint32_t buffers[OPENAL_BUFS];
+    bool source_open;
 } Device;
 
-const char *ddevice_names[2];              /* Default device */
-const char *devices_names[2][MAX_DEVICES]; /* Container of available devices */
-static int size[2];                        /* Size of above containers */
-Device *running[2][MAX_DEVICES] = {{NULL}};     /* Running devices */
-uint32_t primary_device[2];          /* Primary device */
+typedef struct AudioState {
+    ALCdevice *al_device[2];
 
-#ifdef AUDIO
-static ToxAV *av = NULL;
-#endif /* AUDIO */
+    Device devices[2][MAX_DEVICES];
+    uint32_t num_devices[2];
+
+    FrameInfo capture_frame_info;
+    float input_volume;
+
+    // mutexes to prevent changes to input resp. output devices and al_devices
+    // during poll_input iterations resp. calls to write_out;
+    // mutex[input] also used to lock input_volume which poll_input writes to.
+    pthread_mutex_t mutex[2];
+
+    // TODO: unused
+    const char *default_al_device_name[2];              /* Default devices */
 
-/* q_mutex */
-#define lock pthread_mutex_lock(&mutex)
-#define unlock pthread_mutex_unlock(&mutex)
-pthread_mutex_t mutex;
+    const char *al_device_names[2][MAX_OPENAL_DEVICES]; /* Available devices */
+    uint32_t num_al_devices[2];
+    char *current_al_device_name[2];
+} AudioState;
+
+static AudioState *audio_state;
+
+static void lock(DeviceType type)
+{
+    pthread_mutex_lock(&audio_state->mutex[type]);
+}
+
+static void unlock(DeviceType type)
+{
+    pthread_mutex_unlock(&audio_state->mutex[type]);
+}
 
 
-bool thread_running = true,
-     thread_paused = true;               /* Thread control */
+static bool thread_running = true,
+            thread_paused = true;               /* Thread control */
 
-void *thread_poll(void *);
-/* Meet devices */
 #ifdef AUDIO
-DeviceError init_devices(ToxAV *av_)
-#else
+static void *poll_input(void *);
+#endif
+
+static uint32_t sound_mode(bool stereo)
+{
+    return stereo ? AL_FORMAT_STEREO16 : AL_FORMAT_MONO16;
+}
+
+static uint32_t sample_size(bool stereo)
+{
+    return stereo ? 4 : 2;
+}
+
 DeviceError init_devices(void)
-#endif /* AUDIO */
 {
-    get_devices_names();
+    audio_state = calloc(1, sizeof(AudioState));
 
-    // Start poll thread
-    if (pthread_mutex_init(&mutex, NULL) != 0) {
+    if (audio_state == NULL) {
         return de_InternalError;
     }
 
+    get_al_device_names();
+
+    for (DeviceType type = input; type <= output; ++type) {
+        audio_state->al_device[type] = NULL;
+
+        if (pthread_mutex_init(&audio_state->mutex[type], NULL) != 0) {
+            return de_InternalError;
+        }
+    }
+
+#ifdef AUDIO
+    // Start poll thread
     pthread_t thread_id;
 
-    if (pthread_create(&thread_id, NULL, thread_poll, NULL) != 0 || pthread_detach(thread_id) != 0) {
+    if (pthread_create(&thread_id, NULL, poll_input, NULL) != 0 || pthread_detach(thread_id) != 0) {
```

Comment:
This violates MISRA C:2012 required rule 13.5: "The right hand operand of a logical && or || operator shall not
contain persistent side effects". Up to you whether you want to comply, though.